src/control/lib/hardware/hwloc/provider.go

```diff
@@ -336,3 +336,41 @@ func osDevTypeToHardwareDevType(osType int) hardware.DeviceType {
 
 	return hardware.DeviceTypeUnknown
 }
+
+// GetNUMANodeIDForPID fetches the NUMA node ID associated with a given process ID.
+func (p *Provider) GetNUMANodeIDForPID(ctx context.Context, pid int32) (uint, error) {
+	topo, cleanupTopo, err := p.initTopology()
+	if err != nil {
+		return 0, errors.Wrap(err, "initializing topology")
+	}
+	defer cleanupTopo()
+
+	// If there are no NUMA nodes detected, there's nothing to do here.
+	currentNode, err := topo.getNextObjByType(objTypeNUMANode, nil)
+	if err != nil {
+		return 0, hardware.ErrNoNUMANodes
+	}
+
+	cpuSet, cleanupCPUSet, err := topo.getProcessCPUSet(pid, 0)
+	if err != nil {
+		return 0, errors.Wrapf(err, "getting CPU set for PID %d", pid)
+	}
+	defer cleanupCPUSet()
+
+	nodeSet, cleanupNodeSet, err := cpuSet.toNodeSet()
+	if err != nil {
+		return 0, errors.Wrapf(err, "converting PID %d CPU set to NUMA node set", pid)
+	}
+	defer cleanupNodeSet()
```

Comment:
Given that this could be called tens or even in theory hundreds of times a second on a busy compute node, do you foresee any potential performance problems with all of the setup/teardown in here?

Reply:
Good catch. I know that was the motivation behind the netdetect context in the older implementation, which was cached for this call.